tests/src/test/scala/scalaz/MonadErrorTest.scala

```diff
@@ -8,38 +8,46 @@ import Isomorphism.{ <~>, IsoFunctorTemplate }
 object MonadErrorTest extends SpecLite {
 
   trait Decoder[A] {
-    def decode(s: String): String \/ A
+    def decode(s: String): Int \/ A
   }
   object Decoder {
     @inline def apply[A](implicit A: Decoder[A]): Decoder[A] = A
-    @inline def instance[A](f: String => String \/ A): Decoder[A] = new Decoder[A] {
-      override def decode(s: String): String \/ A = f(s)
+    @inline def instance[A](f: String => Int \/ A): Decoder[A] = new Decoder[A] {
+      override def decode(s: String): Int \/ A = f(s)
     }
 
-    // type aliases are needed to help with type inference, even kind-projector
-    // fails us...
-    type Out[a] = String \/ a
-    type MT[a] = ReaderT[Out, String, a]
     implicit val string: Decoder[String] = instance(_.right)
-    implicit val isoReaderT: Decoder <~> MT =
-      new IsoFunctorTemplate[Decoder, MT] {
-        def from[A](fa: MT[A]) = instance(fa.run(_))
-        def to[A](fa: Decoder[A]) = ReaderT[Out, String, A](fa.decode)
-      }
 
-    implicit val monad: MonadError[Decoder, String] = MonadError.fromIso(isoReaderT)
+    // type ascriptions are needed for type inference
+    type I = String
+    type O[a] = Int \/ a
+    type MT[a] = Kleisli[O, I, a]
+    val iso: Decoder <~> MT = Kleisli.iso[Decoder, I, O](
+      instance = λ[λ[a => (I => O[a])] ~> Decoder](instance(_)),
+      decode   = λ[Decoder ~> λ[a => (I => O[a])]](_.decode)
+    )
```

Comment:
Yeah, `Kleisli.iso` helper doesn't really seem like net improvement in boilerplate, just exchanges one kind of boilerplate for another. 😕

Reply:
it is nicer with partial unification, but without partial unification it's pretty much the same amount of typing. The advantage is that at least this boilerplate is concerned only with *your* classes, not the internals of `ReaderT`.

I am tempted to write a `@covariant` annotation to generate this noise.

Also, rewriting Iso to use Pascal sounds :100:

Reply:
Not sure what you mean that `@covariant` annotation would do.

Reply:
Generate the boilerplate that's in this test... the `iso` and the `monad`.

Or perhaps it's something a scalaz alternative to simulacrum could do automatically if it detects a SAM typeclass with the type parameter in a disjunction covariant position.